### PR TITLE
Fix run_options not being used

### PR DESCRIPTION
**Pull request by thisac**
**Context:**
Setting the number of shots in `TDMProgram.run_options` doesn't change the number of shots used when calling `eng.run()`, without passing a shots value.

**Description of the Change:**
Shots now checked and retrieved in the following order: shots passed via `eng.run(prog, shots=shots)` > `prog.run_options` > default value, which is 1. The left-most available being the one that is used.

**Benefits:**
If `TDMProgram.run_options` is set, that value is used rather than falling back on the default value. This can be the case, for example, when converting from a Blackbird script. Without this fix the shots values set in the Blackbird script would simply be ignored when running the engine.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: strawberryfields/engine.py [hunk 1]
```diff
@@ -424,8 +424,9 @@ def run(self, program, *, args=None, compile_options=None, **kwargs):
             not isinstance(program, collections.abc.Sequence) and program.type == "tdm"
         )
         if valid_tdm_program:
+            # priority order for the shots value should be kwargs > run_options > 1
+            shots = kwargs.get("shots", program.run_options.get("shots", 1))
```
Comment (josh146) [inline strawberryfields/engine.py:428]: I like how this use of nested `get` makes it easy to see the priority

**Review by josh146 — APPROVED** [review]
Nice 💪

**Review by antalszava — COMMENTED** [review]
Looking good! 💯 Maybe a test case could be helpful to cover the previous non-working case and verify that shots are extracted well now.

File: tests/frontend/test_engine.py [hunk 2]
```diff
@@ -69,6 +69,46 @@ def test_bad_backend(self):
...
+        eng = sf.Engine("gaussian")
+
+        with prog.context as q:
+            ops.Sgate(0.5) | q[0]
+            ops.MeasureFock() | q
+
+        prog.run_options = {"shots": 5}
+        results = eng.run(prog, shots=2)
+        assert results.samples.shape[0] == 2
+        assert prog.run_options["shots"] == 5
```
Comment (antalszava) [inline tests/frontend/test_engine.py:110]: These tests look awesome! :1st_place_medal: :rocket:

File: tests/frontend/test_tdmprogram.py [hunk 3]
```diff
@@ -706,3 +706,46 @@ def test_move_vac_modes(self, N, crop, expected):
...
+        eng = sf.Engine("gaussian")
+
+        with prog.context([1,2], [3,4]) as (p, q):
+            ops.Sgate(p[0]) | q[0]
+            ops.MeasureHomodyne(p[1]) | q[0]
+
+        prog.run_options = {"shots": 5}
+        results = eng.run(prog, shots=2)
+        assert results.samples.shape[0] == 2
+        assert prog.run_options["shots"] == 5
```
Comment (antalszava) [inline tests/frontend/test_tdmprogram.py:751]: :100:

**Review by antalszava — APPROVED** [review]
Looks good to me! :100:

**Pull request merged**